### PR TITLE
Sarek integration

### PR DESCRIPTION
**Pull request by b97pla**
This PR introduces the new best-practice analysis engine "Sarek". 

Obviously, this is a huge PR so apologies for that. I will try to break it down a bit here and explain the structure of the code. Thanks to the engines being more or less "plugins", there shouldn't be any major impact on the existing pipeline code, so it should be relatively safe to merge.

The overall structure of the Sarek engine module follows e.g. the piper and rnaseq modules, having classes for handling launching, setting up analysis, tracking progress, talking to charon etc. The ambition has been to write code that have separation of concerns, favours short functions and is easy to test. I've also wanted to make it extendable. Whether this is the case remains to be seen but if design choices seem odd or convoluted, this is probably the explanation.

One addition is the "Connectors" (e.g. CharonConnector, ProcessConnector, TrackingConnector etc.) that are used to interface with the various databases and systems. This is to avoid having specific logic about the databases, cluster queues etc. scattered around the code and instead concentrating that to the corresponding connector class.

At this point, the engine can take an organised project and launch the analysis through Slurm. It will record the status in the local tracking database and sync it to Charon. When the update_charon_with_local_jobs_status.py script is run with the "-e sarek" argument, it will check sarek analyses tracked and update Charon accordingly (and remove an entry when analysis finishes). 

Parsing of the results and updating Charon with them is not implemented yet.

### CODE REVIEW

File: ngi_pipeline/engines/sarek/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+class BestPracticeAnalysisNotRecognized(SarekException):
+
+    def __init__(self, best_practice_analysis):
+        super(BestPracticeAnalysisNotRecognized, self).__init__(
+            "best-practice-analysis '{}' not recognized".format(best_practice_analysis))
+        self.best_practice_analysis= best_practice_analysis
+
+
+class ReferenceGenomeNotRecognized(Exception):
```
[the remark this answers is not in the document]
Comment (b97pla) [reply ngi_pipeline/engines/sarek/exceptions.py]: For now, it is intended... The exception handling should probably be refactored at some point though.

File: ngi_pipeline/engines/sarek/models.py [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+    Somatic).
+    """
+
+    DEFAULT_CONFIG = {
+        "profile": "standard",
+        "tools": None,
+        "site_config": None,
+        "sarek_path": "/lupus/ngi/staging/latest/sw/sarek/",
+        "nf_path": "/lupus/ngi/staging/latest/sw/nextflow/nextflow",
+        "containerPath": os.path.join("/lupus/ngi/staging/latest/sw/sarek/", "containers")
```
[the remark this answers is not in the document]
Comment (b97pla) [reply ngi_pipeline/engines/sarek/models.py]: mostly, it's a leftover from testing. if these options are specified in the config on disk, that will override these defaults. I agree though, if they are in the code, they should probably point somewhere meaningful. I'll fix.

**Comment by b97pla** [comment]
Thanks for the review! We could do either way but in the interest of not having pull requests hanging around, I'd vote for merging as well. Just let me fix the configuration issue above and then we should be good to go.

File: ngi_pipeline/engines/sarek/models.py [hunk 3]
```diff
@@ -22,10 +22,10 @@ class SarekAnalysis(object):
...
         "profile": "standard",
         "tools": None,
-        "site_config": None,
-        "sarek_path": "/lupus/ngi/staging/latest/sw/sarek/",
-        "nf_path": "/lupus/ngi/staging/latest/sw/nextflow/nextflow",
-        "containerPath": os.path.join("/lupus/ngi/staging/latest/sw/sarek/", "containers")
+        "config": None,
+        "sarek_path": "/lupus/ngi/production/latest/sw/sarek/",
+        "nf_path": "/lupus/ngi/production/latest/sw/nextflow/nextflow",
+        "containerPath": "/lupus/ngi/staging/latest/resources/singularity/sarek"
```
Comment (senthil10) [inline ngi_pipeline/engines/sarek/models.py]: You left one spot here 😇 (`staging`)

**Comment by senthil10** [comment]
@b97pla Awesome!! Thank you for the change, but looks like you missed a spot to change `staging` -> `production`. I will merge as soon as you fix that :)

**Comment by b97pla** [comment]
@senthil10 my bad, fixed now!

**Pull request merged**